wp/lib/bap_wp/src/z3_utils.ml

```diff
@@ -60,14 +60,44 @@ let mk_smtlib2 (ctx : Z3.context) (smtlib_str : string) (decl_syms : (Z3.FuncDec
   in
   Constr.mk_clause [] goals
 
+let tokenize (str : string) : string list =
+  let delim = Re.Posix.compile_pat "[ \n\r\t()]" in
+  let tokens = Re.split_full delim str in
+  List.rev_map tokens ~f:(function
+      | `Text t -> t
+      | `Delim g ->
+        (* There should always be one value in the group. If not, we will raise an
+           exception. *)
+        if Re.Group.nb_groups g <> 1 then
+          failwith "Number of groups in string delimeter is not 1"
+        else
+          Re.Group.get g 0)
+
+let replace (name : string) (z3_name : string) (tokens : string list) : string list =
+  List.map tokens ~f:(fun token ->
+      if String.equal name token then
+        z3_name
+      else
+        token)
+
 let mk_smtlib2_single (env : Env.t) (smt_post : string) : Constr.t =
   let var_map = Env.get_var_map env in
-  let smt_post = Env.EnvMap.fold var_map ~init:smt_post 
-      ~f:(fun ~key:var ~data:z3_var smt_post ->
-          String.substr_replace_all smt_post ~pattern:((Var.name var) ^ " ") ~with_:((Expr.to_string z3_var) ^ " ")
-        ) in
+  let init_var_map = Env.get_init_var_map env in
```

Comment:
This definitely won't scale to huge smt-lib inputs (you're iterating over the whole s-expression for each name, instead of iterating once over the expression and looking up the var name). But this shouldn't ever matter.

Reply:
You are right! Fixing this should be no problem.